kernel/src/mem.rs

```diff
@@ -66,12 +66,10 @@ pub struct AppSlice<L, T> {
 }
 
 impl<L, T> AppSlice<L, T> {
-    crate fn new(ptr: *mut T, len: usize, appid: AppId) -> AppSlice<L, T> {
-        unsafe {
-            AppSlice {
-                ptr: AppPtr::new(ptr, appid),
-                len: len,
-            }
+    crate fn new(ptr: NonNull<T>, len: usize, appid: AppId) -> AppSlice<L, T> {
```

Comment:
This seems like it should be marked `unsafe`. This constructor can't verify the pointer it receives, and casting a pointer + len to an AppSlice would be bad if the pointer is bad. `unsafe` is the signal that the caller needs to very carefully check before creating an appslice.

Reply:
Agreed. Actually I think there's a more substantial problem here that apps can create non-sync memory aliases, but that's beyond the scope of this PR to fix. I've updated Amit's tracking issue.